File: src/main/org/nlogo/hubnet/mirroring/AgentType.java
// (C) 2012 Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.hubnet.mirroring;

public enum AgentType {
  TURTLE, PATCH, LINK, OBSERVER;

  public int toInt() {
    switch (this) {
      case TURTLE:
        return 0;
      case PATCH:
        return 1;
      case LINK:
        return 2;
      case OBSERVER:
        return 3;
      default:
        throw new IllegalStateException();
    }
  }

  public static AgentType fromAgentClass(Class<? extends org.nlogo.api.Agent> agentClass) {
    if (agentClass == null) {
      return OBSERVER;
    } // don't know why this seems to be necessary - ST 2/25/09
    if (org.nlogo.api.Turtle.class.isAssignableFrom(agentClass)) {
      return TURTLE;
    } else if (org.nlogo.api.Patch.class.isAssignableFrom(agentClass)) {
      return PATCH;
    } else if (org.nlogo.api.Link.class.isAssignableFrom(agentClass)) {
      return LINK;
    } else {
      return OBSERVER;
    }
  }

  public static AgentType fromInt(int type) {
    switch (type) {
      case 0:
        return TURTLE;
      case 1:
        return PATCH;
      case 2:
        return LINK;
      case 3:
        return OBSERVER;
      default:
        throw new IllegalStateException();
    }
  }

}

File: src/main/org/nlogo/hubnet/mirroring/AgentType.scala
// (C) 2012 Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.hubnet.mirroring

import org.nlogo.api

sealed abstract class AgentType(val toInt: Int)

object AgentType {
  case object Turtle extends AgentType(0)
  case object Patch extends AgentType(1)
  case object Link extends AgentType(2)
  case object Observer extends AgentType(3)
  def fromAgentClass(agentClass: Class[_ <: api.Agent]): AgentType =
    if (agentClass == null)
      Observer
    else if (classOf[api.Turtle].isAssignableFrom(agentClass))
      Turtle
    else if (classOf[api.Patch].isAssignableFrom(agentClass))
      Patch
    else if (classOf[api.Link].isAssignableFrom(agentClass))
      Link
    else
      Observer
  def fromInt(i: Int): AgentType = i match {
    case 0 =>
      Turtle
    case 1 =>
      Patch
    case 2 =>
      Link
    case 3 =>
      Observer
  }
}

/**
 * Java can't (I don't think) access Scala inner objects without reflection, so we provide these
 * convenience vals for use from Java.
 */
object AgentTypeJ {
  import AgentType._
  val TURTLE = Turtle
  val PATCH = Patch
  val LINK = Link
  val OBSERVER = Observer
}
